### ISSUE TITLE
7.2.0 won't build in Ubuntu with openjdk-7-java: Directory specified with --with-jdk4 does not exist

### ISSUE THREAD

**Issue by stain** [openlink/virtuoso-opensource#342]
tl;dr: `dpkg-buildpackage` will not complain if `openjdk-7-java` is installed, and `openjdk-6-java` is missing, even though  `configure` falls over without jdk 6. It is not known why both jdk 6 and jdk7 are required.

I started from scratch with the Ubuntu 14.04 x64 Docker image:

```
docker run -it ubuntu:14.04 bash
```

And inside, to build 7.2.0, I did:

```
apt-get update
apt-get install -y unzip wget build-essential
cd /tmp
wget https://github.com/openlink/virtuoso-opensource/archive/stable/7.zip
unzip 7.zip
cd virtuoso-opensource-stable-7/
deps=$(dpkg-checkbuilddeps 2>&1 | sed 's/.*: //' | sed 's/([^)]*)//g')
#  cdbs debhelper quilt autotools-dev autoconf automake libtool flex bison gperf gawk m4 libssl-dev libreadline-dev zlib1g-dev libxml2-dev libpcre3-dev libldap2-dev mono-gmcs cli-common-dev libmono-corlib2.0-cil libmono-system-data2.0-cil libmono-system2.0-cil docbook2x po-debconf libmagickwand-dev java6-sdk java7-jdk
apt-get install -y $deps
```

This fails with

```
Package java6-sdk is a virtual package provided by:
  openjdk-6-jdk 6b34-1.13.6-1ubuntu0.14.04.1
  openjdk-7-jdk 7u75-2.5.4-1~trusty1
  default-jdk 2:1.7-51
You should explicitly select one to install.
```

So I resolved this by doing the mistake of installing just `default-jdk` (e.g. `openjdk-7-jdk`) as I didn't understand why Virtuoso would want both. 

```
apt-get install -y default-jdk
```

Now I raise this bug, as it is likely that here are many Ubuntu 14.04 installations out there that have only `openjdk-7-jdk` (end of life April 2015!) and not the much older `openjdk-6-jdk` - and they won't get this warning as the dependency is "satisfied". 

`dpkg-checkbuilddeps` will now not list java6-sdk anymore:

```
root@becde5ddfd94:/tmp/virtuoso-opensource-stable-7# dpkg-checkbuilddeps 
dpkg-checkbuilddeps: Unmet build dependencies: cdbs debhelper (>= 7) quilt autotools-dev autoconf (>= 2.57) automake (>= 1.10) libtool (>= 1.5.16) flex (>= 2.5.4) bison (>= 1.35) gperf (>= 2.7.2) gawk (>= 3.1.1) m4 (>= 1.4.1) libssl-dev (>= 0.9.7) libreadline-dev zlib1g-dev libxml2-dev libpcre3-dev libldap2-dev mono-gmcs (>= 1.0) cli-common-dev (>= 0.4.4) libmono-corlib2.0-cil (>= 1.2.6) libmono-system-data2.0-cil (>= 1.0) libmono-system2.0-cil (>= 1.2.6) docbook2x po-debconf libmagickwand-dev
```

but building fails:

```
dpkg-buildpackage -us -uc
```

```
...
demo-vad --enable-tutorial-vad --enable-isparql-vad --enable-ods-vad --enable-rdfmappers-vad --enable-sparqldemo-vad --enable-syncml-vad --enable-bpel-vad --with-jdk4= --with-jdk4_1=/usr/lib/jvm/java-7-openjdk-amd64 
...
checking for iODBC location... libsrc/odbcsdk
checking for Java options... done
configure: error: "Directory specified with --with-jdk4 does not exist"
make: *** [debian/stamp-autotools] Error 1
dpkg-buildpackage: error: debian/rules build gave error exit status 2
```

For complete trace, see https://gist.github.com/stain/6ef8b2827d79dfabc460  

As you see `--with-jdk4=` (sic) is set to an empty string in the `configure` phase, while `--with-jdk4_1` has the correct jdk7 path.

I then tried manually to install `openjdk-6-jdk` as well, and tried again. This time virtuoso builds correctly.

```
apt-get install -y openjdk-6-jdk
dpkg-buildpackage -us -uc
```

I am not sure what the fix is (I would not suggest depending on `openjdk-6-jdk` as it is likely to disappear) - but I guess the configure line is what is broken here. Perhaps the first `--with-jdk4` should be dropped.

**Comment by white-gecko** [comment]
I'm experiencing the same issue on debian jessie. Unfortunately there isn't even an `openjdk-6-jdk` package available …

**Comment by spodzone** [comment]
Thanks for reporting this. I'm told the two JDKs support slightly different versions of JDBC, hence the attempt to build both. Will investigate whether we should just drop one.

**Comment by ghirardinicola** [comment]
Any news on this? We have a similar error on debian running this step:

./configure --with-layout=debian --program-transform-name='s/isql$$/isql-vt/;s/isqlw/isqlw-vt/' --with-readline --without-internal-zlib --enable-maintainer-mode --without-iodbc --enable-demo-vad --enable-tutorial-vad --enable-isparql-vad --enable-ods-vad --enable-rdfmappers-vad --enable-sparqldemo-vad --enable-syncml-vad --enable-bpel-vad 

The error is:
cp: cannot stat `../../java.sql.Driver4': No such file or directory 

It compiles on Debian if we remove --with-jdk4 parameters.
What are we missing in that case? jdbc connector?

**Comment by openlink** [comment]
On 31 March 2015 at 14:56, Nicola Ghirardi notifications@github.com wrote:

> ./configure --with-layout=debian
> --program-transform-name='s/isql$$/isql-vt/;s/isqlw/isqlw-vt/'
> --with-readline --without-internal-zlib --enable-maintainer-mode
> --without-iodbc --enable-demo-vad --enable-tutorial-vad
> --enable-isparql-vad --enable-ods-vad --enable-rdfmappers-vad
> --enable-sparqldemo-vad --enable-syncml-vad --enable-bpel-vad
> 
> The error is:
> cp: cannot stat `../../java.sql.Driver4': No such file or directory
> 
> It compiles on Debian if we remove --with-jdk4 parameters.
> What are we missing in that case? jdbc connector?

Hi,

As I thought I'd mentioned, you need both openjdk 6 for the --with-jdk4 and
openjdk 7 for --with-jdk4_1. This is now fixed in
https://github.com/openlink/virtuoso-opensource/commit/d00752fafd4a0cf0eb0dd0734e2512448102424e
.

Regards,

## ~Tim

Tim Haynes
Product Development Consultant
OpenLink Software
http://www.openlinksw.com/
http://twitter.com/openlink

**Comment by joernhees** [comment]
after d00752fafd4a0cf0eb0dd0734e2512448102424e the current develop/7 branch (a4d1005358471ec52b4b11be2b5e21e22ffe20ec) `dpkg-buildpackage -us -uc` seems to work, but only if you don't pass `-j16` to run compilation on multiple cores. Otherwise it seems as if the compilation doesn't wait for dependencies to finish:

```
...
make[5]: Entering directory `/home/hees/virtuoso-opensource/libsrc/JDBCDriverType4/virtuoso/jdbc2'
/usr/bin/cpp -P  -DJDK_VER=17 -Djdbc2=jdbc4 '-DVIRT_JDBC_VER="4.1"' ../../openlink/util/MD5.java >outstd4_1/openlink/util/MD5.java
/usr/bin/cpp -P  -DJDK_VER=17 -Djdbc2=jdbc4 '-DVIRT_JDBC_VER="4.1"' ../../openlink/util/Vector.java >outstd4_1/openlink/util/Vector.java
/usr/bin/cpp -P  -DJDK_VER=17 -Djdbc2=jdbc4 '-DVIRT_JDBC_VER="4.1"' ../../openlink/util/VectorOfDouble.java >outstd4_1/openlink/util/VectorOfDouble.java
/usr/bin/cpp -P  -DJDK_VER=17 -Djdbc2=jdbc4 '-DVIRT_JDBC_VER="4.1"' ../../openlink/util/VectorOfFloat.java >outstd4_1/openlink/util/VectorOfFloat.java
/usr/bin/cpp -P  -DJDK_VER=17 -Djdbc2=jdbc4 '-DVIRT_JDBC_VER="4.1"' ../../openlink/util/VectorOfLong.java >outstd4_1/openlink/util/VectorOfLong.java
cd outstd4_1/virtuoso/sql; cp ../../../../../virtuoso/sql/ExtendedString.java .
cd outstd4_1/virtuoso/sql; cp ../../../../../virtuoso/sql/RdfBox.java .
/usr/bin/cpp -P  -DJDK_VER=17 -Djdbc2=jdbc4 '-DVIRT_JDBC_VER="4.1"' ../../openlink/util/BaseMessage.java >outstd4_1/openlink/util/BaseMessage.java
/bin/bash: outstd4_1/openlink/util/Vector.java: No such file or directory
/bin/bash: outstd4_1/openlink/util/MD5.java: No such file or directory
/usr/bin/cpp -P  -DJDK_VER=17 -Djdbc2=jdbc4 '-DVIRT_JDBC_VER="4.1"' ../../openlink/util/OPLHeapBlob.java >outstd4_1/openlink/util/OPLHeapBlob.java
/bin/bash: outstd4_1/openlink/util/VectorOfFloat.java: No such file or directory
make[5]: *** [outstd4_1/openlink/util/MD5.java] Error 1
make[5]: *** Waiting for unfinished jobs....
...
```

**Comment by openlink** [comment]
I pushed a fix for the make -j6 dependency issue building the JDBC drivers into the latest develop/7 tree.

**Comment by joernhees** [comment]
:+1: thanks

**Comment by white-gecko** [comment]
How can I build a debian package with this fix? Just removing `openjdk-6-jdk` dependency from `debian/control` doesn't fix the package building.

**Comment by joernhees** [comment]
@white-gecko as Tim said above, at the moment you need both openjdk-6 and openjdk-7 to build the 2 slightly different JDBC drivers for JDBC versions 4.0 (jdk6) and JDBC version 4.1 (jdk7).

I think the reported `--with-jdk4` and `--with-jdk4_1` options are a confusing misnomer in virtuoso's configure scripts and should instead either be called `--with-jdk6` and `--with-jdk7` or `--with-jdbc4` and `--with-jdbc4_1`.

@white-gecko As a workaround for now: If debian jessie doesn't provide a jdk6 package out of the box, you can probably just use one from wheezy.

@openlink It would be cool if you guys removed openjdk-6 from the dependencies (and only left openjdk-7 as a hard dependency in there). Then later you could auto-detect if openjdk-6 is still present and only if so automagially add the `--with-jdk4`/`--with-jdk6`/`--with-jdbc4` (however you end up naming it) flag by default. That way the package would build on Ubuntu and Debian.

**Comment by white-gecko** [comment]
The changes provided in PR #404 fix the package building for debian jessie
